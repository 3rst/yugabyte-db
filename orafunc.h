#ifndef __ORAFUNC__
#define __ORAFUNC__

#include "postgres.h"
#include "catalog/catversion.h"
#include "nodes/pg_list.h"
#include <sys/time.h>
#include "utils/datetime.h"

#define TextPGetCString(t) \
        DatumGetCString(DirectFunctionCall1(textout, PointerGetDatum(t))) 
#define CStringGetTextP(c) \
        DatumGetTextP(DirectFunctionCall1(textin, CStringGetDatum(c)))


text* ora_substr(text *str, int start, int len, bool valid_length);
text* ora_make_text_fix(char *c, int n);
int   ora_instr(text *txt, text *pattern, int start, int nth);
text* ora_clone_text(text *t);
int ora_mb_strlen(text *str, char **sizes, int **positions);
int ora_mb_strlen1(text *str);

#endif
