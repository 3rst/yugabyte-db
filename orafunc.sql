SET search_path = public;

CREATE OR REPLACE FUNCTION next_day(value date, weekday text) 
RETURNS date
AS '$libdir/orafunc'
LANGUAGE 'C' IMMUTABLE STRICT;

CREATE OR REPLACE FUNCTION last_day(value date) 
RETURNS date
AS '$libdir/orafunc'
LANGUAGE 'C' IMMUTABLE STRICT;

CREATE OR REPLACE FUNCTION months_between(date1 date, date2 date) 
RETURNS float8
AS '$libdir/orafunc'
LANGUAGE 'C' IMMUTABLE STRICT;

CREATE OR REPLACE FUNCTION add_months(day date, value int) 
RETURNS date
AS '$libdir/orafunc'
LANGUAGE 'C' IMMUTABLE STRICT;
