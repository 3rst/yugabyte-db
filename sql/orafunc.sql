--
-- test built-in date type oracle compatibility functions
--

SELECT add_months ('2003-08-01', 3);
SELECT add_months ('2003-08-01', -3);
SELECT add_months ('2003-08-21', -3);
SELECT add_months ('2003-01-31', 1);

SELECT last_day (to_date('2003/03/15', 'yyyy/mm/dd'));
SELECT last_day (to_date('2003/02/03', 'yyyy/mm/dd'));
SELECT last_day (to_date('2004/02/03', 'yyyy/mm/dd'));

SELECT next_day ('2003-08-01', 'TUESDAY');
SELECT next_day ('2003-08-06', 'WEDNESDAY');
SELECT next_day ('2003-08-06', 'SUNDAY');

SELECT months_between (to_date ('2003/01/01', 'yyyy/mm/dd'), to_date ('2003/03/14', 'yyyy/mm/dd'));
SELECT months_between (to_date ('2003/07/01', 'yyyy/mm/dd'), to_date ('2003/03/14', 'yyyy/mm/dd'));
SELECT months_between (to_date ('2003/07/02', 'yyyy/mm/dd'), to_date ('2003/07/02', 'yyyy/mm/dd'));
SELECT months_between (to_date ('2003/08/02', 'yyyy/mm/dd'), to_date ('2003/06/02', 'yyyy/mm/dd'));
