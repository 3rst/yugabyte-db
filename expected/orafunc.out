--
-- test built-in date type oracle compatibility functions
--
SELECT add_months ('2003-08-01', 3);
 add_months 
------------
 2003-11-01
(1 row)

SELECT add_months ('2003-08-01', -3);
 add_months 
------------
 2003-05-01
(1 row)

SELECT add_months ('2003-08-21', -3);
 add_months 
------------
 2003-05-21
(1 row)

SELECT add_months ('2003-01-31', 1);
 add_months 
------------
 2003-02-28
(1 row)

SELECT last_day (to_date('2003/03/15', 'yyyy/mm/dd'));
  last_day  
------------
 2003-03-31
(1 row)

SELECT last_day (to_date('2003/02/03', 'yyyy/mm/dd'));
  last_day  
------------
 2003-02-28
(1 row)

SELECT last_day (to_date('2004/02/03', 'yyyy/mm/dd'));
  last_day  
------------
 2004-02-29
(1 row)

SELECT next_day ('2003-08-01', 'TUESDAY');
  next_day  
------------
 2003-08-05
(1 row)

SELECT next_day ('2003-08-06', 'WEDNESDAY');
  next_day  
------------
 2003-08-13
(1 row)

SELECT next_day ('2003-08-06', 'SUNDAY');
  next_day  
------------
 2003-08-10
(1 row)

SELECT months_between (to_date ('2003/01/01', 'yyyy/mm/dd'), to_date ('2003/03/14', 'yyyy/mm/dd'));
  months_between   
-------------------
 -2.41935483870968
(1 row)

SELECT months_between (to_date ('2003/07/01', 'yyyy/mm/dd'), to_date ('2003/03/14', 'yyyy/mm/dd'));
  months_between  
------------------
 3.58064516129032
(1 row)

SELECT months_between (to_date ('2003/07/02', 'yyyy/mm/dd'), to_date ('2003/07/02', 'yyyy/mm/dd'));
 months_between 
----------------
              0
(1 row)

SELECT months_between (to_date ('2003/08/02', 'yyyy/mm/dd'), to_date ('2003/06/02', 'yyyy/mm/dd'));
 months_between 
----------------
              2
(1 row)

